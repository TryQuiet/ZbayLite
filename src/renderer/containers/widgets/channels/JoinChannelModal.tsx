import React from 'react'
import { useSelector, useDispatch } from 'react-redux'
import { publicChannels } from '@zbayapp/nectar'
import JoinChannelModalComponent from '../../../components/widgets/channels/JoinChannelModal'
import channelHandlers from '../../../store/handlers/channel'
// import notificationsHandlers from '../../../store/handlers/notifications'
import modalsHandlers from '../../../store/handlers/modals'
import modalsSelectors from '../../../store/selectors/modals'

const useJoinChannelData = () => {
  const modalName = 'joinChannel'
  const data = {
    publicChannels: useSelector(publicChannels.selectors.publicChannels),
    users: [],
    open: useSelector(modalsSelectors.open(modalName)),
    modalName
  }
  return data
}

export const useJoinChannelActions = () => {
  const dispatch = useDispatch()
  const modalName = 'joinChannel'
  const actions = {
    joinChannel: channel => {
      dispatch(channelHandlers.epics.linkChannelRedirect(channel))
    },
    handleClose: () => {
      dispatch(modalsHandlers.closeModalHandler(modalName))
    }
    // showNotification: ()  => {dispatch(notificationsHandlers.actions.enqueueSnackbar()}
  }
  return actions
}

export const JoinChannelModal = () => {
  const { publicChannels, open, users } = useJoinChannelData()
  const { joinChannel, handleClose } = useJoinChannelActions()

  return (
    <JoinChannelModalComponent
      publicChannels={publicChannels}
      joinChannel={joinChannel}
      // showNotification={showNotification}
      open={open}
      users={users}
      handleClose={handleClose}
    />
  )
}

export default JoinChannelModal
