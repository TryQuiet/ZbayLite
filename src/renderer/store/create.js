import { applyMiddleware, createStore, compose } from 'redux'
import thunk from 'redux-thunk'
import promise from 'redux-promise-middleware'
import createDebounce from 'redux-debounced'

import reducers from './reducers'
import { errorsMiddleware } from './middlewares'

const composer = (enhancers) => {
  if (window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ &&
      process.env.NODE_ENV === 'development') {
    return window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__(enhancers)
  }
  return compose(enhancers)
}

export default ({
  initialState
} = {}) => createStore(
  reducers,
  initialState,
  composer(applyMiddleware(...[errorsMiddleware, createDebounce(), thunk, promise()]))
)
