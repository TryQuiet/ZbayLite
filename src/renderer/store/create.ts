import { applyMiddleware, createStore } from 'redux'
import thunk from 'redux-thunk'
import promise from 'redux-promise-middleware'
import createDebounce from 'redux-debounced'
import createSagaMiddleware from 'redux-saga'
import rootSaga from '../sagas/index.saga'
import { composeWithDevTools } from 'redux-devtools-extension'

import reducers from './reducers'
import { errorsMiddleware } from './middlewares'

const sagaMiddleware = createSagaMiddleware()

export default (initialState = {}) => {
  const store = createStore(
    reducers,
    initialState,
    composeWithDevTools(
      applyMiddleware(...[errorsMiddleware, createDebounce(), sagaMiddleware, thunk, promise()])
    )
  )
  sagaMiddleware.run(rootSaga)
  return store
}
