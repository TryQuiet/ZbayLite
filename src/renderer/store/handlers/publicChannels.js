import Immutable from 'immutable'
import { createAction, handleActions } from 'redux-actions'
import * as R from 'ramda'
import feesSelectors from '../selectors/fees'
import nodeSelectors from '../selectors/node'
import client from '../../zcash'
import { errorNotification, successNotification } from './utils'
import identitySelectors from '../selectors/identity'
import notificationsHandlers from './notifications'
import messagesOperators from '../../zbay/messages'
import { ADDRESS_TYPE } from '../../zbay/transit'
import feesHandlers from '../handlers/fees'
import staticChannels from '../../zcash/channels'
import { messageType, actionTypes } from '../../../shared/static'
import { checkTransferCount } from '../handlers/messages'

export const _PublicChannelData = Immutable.Record(
  {
    address: '',
    name: '',
    description: '',
    owner: '',
    timestamp: 0,
    keys: {}
  },
  'PublicChannelData'
)
export const initialState = Immutable.Map({})

export const setPublicChannels = createAction(actionTypes.SET_PUBLIC_CHANNELS)

export const actions = {
  setPublicChannels
}
export const fetchPublicChannels = (address, messages) => async (dispatch, getState) => {
  try {
    const transferCountFlag = await dispatch(
      checkTransferCount(address, messages)
    )
    if (transferCountFlag === -1 || !messages) {
      return
    }
    const filteredZbayMessages = messages.filter(msg =>
      msg.memohex.startsWith('ff')
    )

    const registrationMessages = await Promise.all(
      filteredZbayMessages.map(transfer => {
        const message = messagesOperators.transferToMessage(transfer)
        return message
      })
    )
    const sortedMessages = registrationMessages
      .filter(msg => msg !== null)
    let minfee = 0
    let publicChannelsMap = Immutable.Map({})
    const network = nodeSelectors.network(getState())
    for (const msg of sortedMessages) {
      if (
        msg.type === messageType.CHANNEL_SETTINGS &&
        staticChannels.zbay[network].publicKey === msg.publicKey
      ) {
        minfee = parseFloat(msg.message.minFee)
      }
      if (!msg.spent.gte(minfee) || msg.type !== messageType.PUBLISH_CHANNEL) {
        continue
      }
      const updateChannelSettings = R.findLast(
        settingsMsg =>
          settingsMsg.type === messageType.CHANNEL_SETTINGS_UPDATE &&
          settingsMsg.publicKey === msg.publicKey &&
          settingsMsg.message.updateChannelAddress ===
            msg.message.channelAddress
      )(sortedMessages)
      const channel = _PublicChannelData({
        address: msg.message.channelAddress,
        name: msg.message.channelName,
        description: updateChannelSettings
          ? updateChannelSettings.message.updateChannelDescription
          : msg.message.channelDescription,
        owner: msg.publicKey,
        keys: { ivk: msg.message.channelIvk },
        timestamp: msg.createdAt
      })
      if (channel !== null && !publicChannelsMap.get(channel.name)) {
        publicChannelsMap = publicChannelsMap.merge({ [channel.name]: channel })
      }
    }
    await dispatch(feesHandlers.actions.setPublicChannelFee(minfee))
    await dispatch(setPublicChannels(publicChannelsMap))
  } catch (err) {
    console.warn(err)
  }
}
export const publishChannel = ({
  channelAddress,
  channelName,
  channelDescription,
  channelIvk
}) => async (dispatch, getState) => {
  const identityAddress = identitySelectors.address(getState())
  const privKey = identitySelectors.signerPrivKey(getState())
  const network = nodeSelectors.network(getState())
  const fee = feesSelectors.publicChannelfee(getState())
  const message = messagesOperators.createMessage({
    messageData: {
      type: messagesOperators.messageType.PUBLISH_CHANNEL,
      data: {
        channelName,
        channelAddress,
        channelIvk,
        channelDescription,
        networkType:
          network === 'testnet'
            ? ADDRESS_TYPE.SHIELDED_TESTNET
            : ADDRESS_TYPE.SHIELDED_MAINNET
      }
    },
    privKey: privKey
  })
  const transfer = await messagesOperators.messageToTransfer({
    message,
    address: staticChannels.channelOfChannels[network].address,
    identityAddress,
    amount: fee
  })
  try {
    const txid = await client.sendTransaction(transfer)
    if (txid.error) {
      throw new Error(txid.error)
    }
    dispatch(
      notificationsHandlers.actions.enqueueSnackbar(
        successNotification({
          message: 'Your channel has been published.'
        })
      )
    )
  } catch (err) {
    dispatch(
      notificationsHandlers.actions.enqueueSnackbar(
        errorNotification({
          message: "Couldn't create channel, please check node connection."
        })
      )
    )
  }
}
export const epics = {
  fetchPublicChannels,
  publishChannel
}

export const reducer = handleActions(
  {
    [setPublicChannels]: (state, { payload: publicChannels }) => {
      return state.merge(publicChannels)
    }
  },
  initialState
)

export default {
  reducer,
  epics,
  actions
}
