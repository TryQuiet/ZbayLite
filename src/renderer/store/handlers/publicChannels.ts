import { produce, immerable } from 'immer'
import { createAction, handleActions } from 'redux-actions'
import * as R from 'ramda'
import feesSelectors from '../selectors/fees'
import nodeSelectors from '../selectors/node'
import client from '../../zcash'
import { errorNotification, successNotification } from './utils'
import identitySelectors from '../selectors/identity'
import notificationsHandlers from './notifications'
import messagesOperators from '../../zbay/messages'
import { ADDRESS_TYPE } from '../../zbay/transit'
import feesHandlers from './fees'
import staticChannels from '../../zcash/channels'
import { messageType, actionTypes } from '../../../shared/static'

import { ActionsType, PayloadType } from './types'

import { DisplayableMessage } from '../../zbay/messages.types'

// Used only in some tests
export const _PublicChannelData = {
  address: '',
  name: '',
  description: '',
  owner: '',
  timestamp: 0,
  keys: {}
}

export class PublicChannel {
  address: string
  name: string
  description: string
  owner: string
  keys: { ivk?: string, sk?: string }
  timestamp: number

  constructor(values: Partial<PublicChannel>) {
    Object.assign(this, values)
    this[immerable] = true
  }
}

export type PublicChannelsStore = { [name: string]: PublicChannel }

export const initialState: PublicChannelsStore = {}

export const setPublicChannels = createAction<{ [name: string]: PublicChannel }>(
  actionTypes.SET_PUBLIC_CHANNELS
)

export const actions = {
  setPublicChannels
}

export type PublicChannelsActions = ActionsType<typeof actions>

export const fetchPublicChannels = (address, messages: DisplayableMessage[]) => async (
  dispatch,
  getState
) => {
  try {
    const filteredZbayMessages = messages.filter(msg => msg.memohex.startsWith('ff'))

    const registrationMessages: DisplayableMessage[] = await Promise.all(
      filteredZbayMessages.map(transfer => {
        const message = messagesOperators.transferToMessage(transfer)
        return message
      })
    )
    const sortedMessages = registrationMessages.filter(msg => msg !== null)
    let minfee = 0
    let publicChannelsMap = {}
    const network = nodeSelectors.network(getState())
    for (const msg of sortedMessages) {
      if (
        msg.type === messageType.CHANNEL_SETTINGS &&
        staticChannels.zbay[network].publicKey === msg.publicKey
      ) {
        minfee = parseFloat(msg.message.minFee)
      }
      if (!msg.spent.gte(minfee) || msg.type !== messageType.PUBLISH_CHANNEL) {
        continue
      }
      const updateChannelSettings = R.findLast(
        (settingsMsg: DisplayableMessage) =>
          settingsMsg.type === messageType.CHANNEL_SETTINGS_UPDATE &&
          settingsMsg.publicKey === msg.publicKey &&
          settingsMsg.message.updateChannelAddress === msg.message.channelAddress
      )(sortedMessages)
      const channel = new PublicChannel({
        address: msg.message.channelAddress,
        name: msg.message.channelName,
        description: updateChannelSettings
          ? updateChannelSettings.message.updateChannelDescription
          : msg.message.channelDescription,
        owner: msg.publicKey,
        keys: { ivk: msg.message.channelIvk },
        timestamp: msg.createdAt
      })
      if (channel !== null && !publicChannelsMap[channel.name]) {
        publicChannelsMap = {
          ...publicChannelsMap,
          [channel.name]: {
            ...channel
          }
        }
      }
    }
    await dispatch(feesHandlers.actions.setPublicChannelFee(minfee))
    console.log('tes public channelst', publicChannelsMap)
    await dispatch(setPublicChannels(publicChannelsMap))
  } catch (err) {
    console.warn(err)
  }
}
export const publishChannel = ({
  channelAddress,
  channelName,
  channelDescription,
  channelIvk
}) => async (dispatch, getState) => {
  const identityAddress = identitySelectors.address(getState())
  const privKey = identitySelectors.signerPrivKey(getState())
  const network = nodeSelectors.network(getState())
  const fee = feesSelectors.publicChannelfee(getState())
  const message = messagesOperators.createMessage({
    messageData: {
      type: messagesOperators.messageType.PUBLISH_CHANNEL,
      data: {
        channelName,
        channelAddress,
        channelIvk,
        channelDescription,
        networkType:
          network === 'testnet' ? ADDRESS_TYPE.SHIELDED_TESTNET : ADDRESS_TYPE.SHIELDED_MAINNET
      }
    },
    privKey: privKey
  })
  const transfer = await messagesOperators.messageToTransfer({
    message,
    address: staticChannels.channelOfChannels[network].address,
    identityAddress,
    amount: fee
  })
  try {
    const txid = await client.sendTransaction(transfer)
    if (txid.error) {
      throw new Error(txid.error)
    }
    dispatch(
      notificationsHandlers.actions.enqueueSnackbar(
        successNotification({
          message: 'Your channel has been published.'
        })
      )
    )
  } catch (err) {
    dispatch(
      notificationsHandlers.actions.enqueueSnackbar(
        errorNotification({
          message: "Couldn't create channel, please check node connection."
        })
      )
    )
  }
}
export const epics = {
  fetchPublicChannels,
  publishChannel
}

export const reducer = handleActions<PublicChannelsStore, PayloadType<PublicChannelsActions>>(
  {
    [setPublicChannels.toString()]: (
      state,
      { payload: publicChannels }: PublicChannelsActions['setPublicChannels']
    ) =>
      produce(state, draft => {
        return {
          ...draft,
          ...publicChannels
        }
      })
  },
  initialState
)

export default {
  reducer,
  epics,
  actions
}
