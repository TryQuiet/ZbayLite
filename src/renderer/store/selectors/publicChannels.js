import { createSelector } from 'reselect'

const store = s => s

const publicChannels = createSelector(
  store,
  state => state.get('publicChannels')
)

const publicChannelsByAddress = address =>
  createSelector(
    publicChannels,
    channels => channels[address]
  )

export default {
  publicChannelsByAddress,
  publicChannels
}
