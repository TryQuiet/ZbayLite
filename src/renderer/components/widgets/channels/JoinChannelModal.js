import React from 'react'
import PropTypes from 'prop-types'
import { DateTime } from 'luxon'
import * as R from 'ramda'
import { Formik, Form } from 'formik'
import { Typography } from '@material-ui/core'
import Grid from '@material-ui/core/Grid'
import { withStyles } from '@material-ui/core/styles'
import TextField from '@material-ui/core/TextField'

import Modal from '../../ui/Modal/Modal'
import { AutocompleteField } from '../../ui/Autocomplete/Autocomplete'
import { errorNotification } from '../../../store/handlers/utils'
import LoadindButton from '../../ui/LoadingButton/LoadingButton'

const styles = theme => ({
  root: {
    padding: 32,
    height: '100%',
    width: '100%'
  },
  fullContainer: {
    width: '100%',
    height: '100%'
  },
  input: {
    marginBottom: 0,
    marginTop: 0,
    width: '100%'
  },
  button: {
    width: 151,
    height: 60,
    fontSize: 16,
    backgroundColor: theme.palette.colors.zbayBlue,
    color: theme.palette.colors.white,
    '&:hover': {
      backgroundColor: theme.palette.colors.zbayBlue
    }
  },
  title: {
    marginBottom: 24
  },
  info: {
    marginTop: 8,
    lineHeight: '18px',
    color: theme.palette.colors.darkGray,
    letterSpacing: '0.4px'
  },
  channelTitle: {
    lineHeight: '26px',
    fontSize: 16
  },
  channelInfo: {
    lineHeight: '18px',
    color: theme.palette.colors.darkGray,
    letterSpacing: '0.4px'
  },
  option: {},
  materialOption: {
    height: 68
  },
  timeInfo: {
    marginTop: -16,
    lineHeight: '18px',
    color: theme.palette.colors.darkGray,
    letterSpacing: '0.4px',
    marginBottom: 16
  },
  description: {
    fontSize: 14,
    lineHeight: '24px'
  },
  informationBox: {
    backgroundColor: theme.palette.colors.gray03,
    borderRadius: 4,
    height: 42,
    color: theme.palette.colors.trueBlack,
    letterSpacing: '0.4px',
    lineHeight: '18px',
    fontSize: 12,
    paddingLeft: 16,
    paddingRight: 16,
    marginBottom: 24,
    marginTop: 16
  }
})
export const JoinChannelModal = ({
  classes,
  open,
  handleClose,
  joinChannel,
  Channel,
  publicChannels,
  showNotification,
  users
}) => {
  const channelsArray = Array.from(Object.values(publicChannels))
  const [step, setStep] = React.useState(0)
  const [loading, setLoading] = React.useState(false)
  return (
    <Modal
      open={open}
      handleClose={() => {
        handleClose()
        setStep(0)
      }}
      title=''
      fullPage>
      <Grid className={classes.root}>
        <Formik
          onSubmit={async (values, { resetForm }) => {
            const ch = channelsArray.find(channel => channel.name === values.channel.name)
            if (ch) {
              if (!ch) {
                showNotification(errorNotification({ message: 'Channel does not exist' }))
                return
              }
              setLoading(true)
              await joinChannel(ch)
              setLoading(false)
              setStep(0)
              handleClose()
              resetForm()
              return
            }
            showNotification(
              errorNotification({
                message: 'There was an error. Please check channel URL'
              })
            )
          }}>
          {({ values, setFieldValue }) => {
            return (
              <Form className={classes.fullContainer}>
                <Grid
                  container
                  justify='flex-start'
                  direction='column'
                  className={classes.fullContainer}>
                  <Typography variant='h3' className={classes.title}>
                    {step === 0 ? 'Search for Channels' : `#${values.channel.name}`}
                  </Typography>
                  {step !== 0 && (
                    <Typography variant='caption' className={classes.timeInfo}>
                      {`Created by @${
                        users[values.channel.owner]
                          ? users[values.channel.owner].nickname
                          : 'Unnamed'
                      } on ${DateTime.fromSeconds(parseInt(values.channel.timestamp)).toFormat(
                        'LLL d, y'
                      )} `}
                    </Typography>
                  )}
                  {step === 0 ? (
                    <AutocompleteField
                      name={'channel'}
                      classes={{ option: classes.materialOption }}
                      options={channelsArray}
                      renderOption={option => {
                        const time = DateTime.fromSeconds(parseInt(option.timestamp))
                        return (
                          <Grid container direction='column' className={classes.option}>
                            <Typography variant='body1' className={classes.channelTitle}>
                              {`#${option.name}`}
                            </Typography>
                            <Typography variant='caption' className={classes.channelInfo}>
                              {`Created by @${
                                users[option.owner] ? users[option.owner].nickname : 'Unnamed'
                              } on ${time.toFormat('LLL d, y')} `}
                            </Typography>
                          </Grid>
                        )
                      }}
                      value={values.channel}
                      onChange={(e, v) => {
                        setFieldValue('channel', v)
                        setStep(1)
                      }}
                      renderInput={params => (
                        <TextField
                          {...params}
                          className={classes.input}
                          variant='outlined'
                          placeholder={'Search'}
                          margin='normal'
                        />
                      )}
                    />
                  ) : (
                    <>
                      <Typography variant='body2' className={classes.description}>
                        {`${values.channel.description}`}
                      </Typography>
                      <Grid container alignItems='center' className={classes.informationBox}>
                        After joining, it may take some time for messages to fully load.
                      </Grid>
                    </>
                  )}

                  {step !== 0 ? (
                    <LoadindButton
                      variant='contained'
                      color='primary'
                      size='large'
                      type='submit'
                      text='Join Channel'
                      classes={{ button: classes.button }}
                      inProgress={loading}
                      disabled={loading}
                    />
                  ) : (
                    <Typography variant='caption' className={classes.info}>
                      If you have an invite link, open it in a browser
                    </Typography>
                  )}
                </Grid>
              </Form>
            )
          }}
        </Formik>
      </Grid>
    </Modal>
  )
}

JoinChannelModal.propTypes = {
  classes: PropTypes.object.isRequired,
  open: PropTypes.bool.isRequired,
  handleClose: PropTypes.func.isRequired,
  joinChannel: PropTypes.func.isRequired,
  showNotification: PropTypes.func.isRequired,
  publicChannels: PropTypes.object.isRequired,
  users: PropTypes.object.isRequired
}

export default R.compose(React.memo, withStyles(styles))(JoinChannelModal)
