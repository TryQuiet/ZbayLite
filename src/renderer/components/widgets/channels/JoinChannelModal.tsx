import React from 'react'
import { DateTime } from 'luxon'
import { Formik, Form } from 'formik'
import { Typography } from '@material-ui/core'
import Grid from '@material-ui/core/Grid'
import { makeStyles } from '@material-ui/core/styles'
import TextField from '@material-ui/core/TextField'

import Modal from '../../ui/Modal/Modal'
import { AutocompleteField } from '../../ui/Autocomplete/Autocomplete'
import { errorNotification, IErrorNotification } from '../../../store/handlers/utils'
import LoadindButton from '../../ui/LoadingButton/LoadingButton'
import { Channel } from '../../../store/handlers/channel'
import { User } from '@zbayapp/nectar/lib/sagas/users/users.slice'

const useStyles = makeStyles((theme) => ({
  root: {
    padding: 32,
    height: '100%',
    width: '100%'
  },
  fullContainer: {
    width: '100%',
    height: '100%'
  },
  input: {
    marginBottom: 0,
    marginTop: 0,
    width: '100%'
  },
  button: {
    width: 151,
    height: 60,
    fontSize: 16,
    backgroundColor: theme.palette.colors.zbayBlue,
    color: theme.palette.colors.white,
    '&:hover': {
      backgroundColor: theme.palette.colors.zbayBlue
    }
  },
  title: {
    marginBottom: 24
  },
  info: {
    marginTop: 8,
    lineHeight: '18px',
    color: theme.palette.colors.darkGray,
    letterSpacing: '0.4px'
  },
  channelTitle: {
    lineHeight: '26px',
    fontSize: 16
  },
  channelInfo: {
    lineHeight: '18px',
    color: theme.palette.colors.darkGray,
    letterSpacing: '0.4px'
  },
  option: {},
  materialOption: {
    height: 68
  },
  timeInfo: {
    marginTop: -16,
    lineHeight: '18px',
    color: theme.palette.colors.darkGray,
    letterSpacing: '0.4px',
    marginBottom: 16
  },
  description: {
    fontSize: 14,
    lineHeight: '24px'
  },
  informationBox: {
    backgroundColor: theme.palette.colors.gray03,
    borderRadius: 4,
    height: 42,
    color: theme.palette.colors.trueBlack,
    letterSpacing: '0.4px',
    lineHeight: '18px',
    fontSize: 12,
    paddingLeft: 16,
    paddingRight: 16,
    marginBottom: 24,
    marginTop: 16
  }
}))

interface JoinChannelModalProps {
  open: boolean
  handleClose: () => void
  joinChannel: () => void
  publicChannels: Channel
  showNotification: (arg: IErrorNotification) => void
  users: User
}

export const JoinChannelModal: React.FC<JoinChannelModalProps> = ({
  open,
  handleClose,
  joinChannel,
  publicChannels,
  showNotification,
  users
}) => {
  const classes = useStyles({})
  const channelsArray = Array.from(Object.values(publicChannels))
  const [step, setStep] = React.useState(0)
  const [loading, setLoading] = React.useState(false)
  return (
    <Modal
      open={open}
      handleClose={() => {
        handleClose()
        setStep(0)
      }}
      title=''
      fullPage>
      <Grid className={classes.root}>
        <Formik
          initialValues={{
            channel: {
              name: '',
              owner: '',
              timestamp: '',
              description: ''
            }
          }}
          onSubmit={async (values, { resetForm }) => {
            const ch = channelsArray.find(channel => channel.name === values.channel.name)
            if (ch) {
              if (!ch) {
                showNotification(errorNotification({ message: 'Channel does not exist' }))
                return
              }
              setLoading(true)
              await joinChannel()
              setLoading(false)
              setStep(0)
              handleClose()
              resetForm()
              return
            }
            showNotification(
              errorNotification({
                message: 'There was an error. Please check channel URL'
              })
            )
          }}>
          {({ values, setFieldValue }) => {
            return (
              <Form className={classes.fullContainer}>
                <Grid
                  container
                  justify='flex-start'
                  direction='column'
                  className={classes.fullContainer}>
                  <Typography variant='h3' className={classes.title}>
                    {step === 0 ? 'Search for Channels' : `#${values.channel.name}`}
                  </Typography>
                  {step !== 0 && (
                    <Typography variant='caption' className={classes.timeInfo}>
                      {`Created by @${users[values.channel.owner]
                        ? users[values.channel.owner].nickname
                        : 'Unnamed'
                        } on ${DateTime.fromSeconds(
                          parseInt(values.channel.timestamp)
                        ).toFormat('LLL d, y')} `}
                    </Typography>
                  )
                  }
                  {step === 0 ? (
                    <AutocompleteField
                      name={'channel'}
                      classes={{ option: classes.materialOption }}
                      options={channelsArray}
                      renderOption={option => {
                        const time = DateTime.fromSeconds(parseInt(option.timestamp))
                        return (
                          <Grid container direction='column' className={classes.option}>
                            <Typography variant='body1' className={classes.channelTitle}>
                              {`#${option.name}`}
                            </Typography>
                            <Typography
                              variant='caption'
                              className={classes.channelInfo}
                            >
                              {`Created by @${users[option.owner]
                                ? users[option.owner].nickname
                                : 'Unnamed'
                                } on ${time.toFormat('LLL d, y')} `}
                            </Typography>
                          </Grid>
                        )
                      }}
                      value={values.channel}
                      onChange={(v: string) => {
                        setFieldValue('channel', v)
                        setStep(1)
                      }}
                      renderInput={params => (
                        <TextField
                          {...params}
                          className={classes.input}
                          variant='outlined'
                          placeholder={'Search'}
                          margin='normal'
                        />
                      )}
                    />
                  ) : (
                    <>
                      <Typography variant='body2' className={classes.description}>
                        {`${values.channel.description}`}
                      </Typography>
                      <Grid container alignItems='center' className={classes.informationBox}>
                        After joining, it may take some time for messages to fully load.
                      </Grid>
                    </>
                  )}

                  {step !== 0 ? (
                    <LoadindButton
                      variant='contained'
                      color='primary'
                      size='large'
                      type='submit'
                      text='Join Channel'
                      classes={{ button: classes.button }}
                      inProgress={loading}
                      disabled={loading}
                    />
                  ) : (
                    <Typography variant='caption' className={classes.info}>
                      If you have an invite link, open it in a browser
                    </Typography>
                  )}
                </Grid>
              </Form>
            )
          }}
        </Formik >
      </Grid >
    </Modal >
  )
}

export default JoinChannelModal
