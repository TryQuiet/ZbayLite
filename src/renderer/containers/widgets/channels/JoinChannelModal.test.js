/* eslint import/first: 0 */
import { mapStateToProps, mapDispatchToProps } from './JoinChannelModal'

import create from '../../../store/create'
import { _PublicChannelData } from '../../../store/handlers/publicChannels'

describe('JoinChannelModal', () => {
  let store = null
  beforeEach(() => {
    jest.clearAllMocks()
    store = create({
      initialState: {
        importedChannel: {
          General: {
            ..._PublicChannelData,
            address:
              'ztestsapling1dfjv308amnk40s89trkvz646ne69553us0g858mmpgsw540efgftn4tf25gts2cttg3jgk9y8lx',
            minFee: '0',
            name: 'General',
            description: 'This is a general channel available to all users of Zbay by default.',
            onlyForRegistered: '0',
            owner: '0208be86d3cac41fdb539b0b761bedccedaa300d5a09fd3ca34b6acad1ba856bcb'
          },
          Test: {
            ..._PublicChannelData,
            address:
              'ztestsapling1x7wn5g6y3c9fjnv0k78ks7dfprpuk0uvqmjxye0pwnwf73yh50krkgyempp09fjdlzaxuz90wxx',
            minFee: '0',
            name: 'Test',
            description: '22222',
            onlyForRegistered: '0',
            owner: '0208be86d3cac41fdb539b0b761bedccedaa300d5a09fd3ca34b6acad1ba856bcb',
            keys: {
              ivk: 'zivktestsapling15seuz0zraqvkhrc9elhm4k4xvnsgmyy4tkfka8lm5j62z4hejgzq3n3xxw'
            }
          }
        }
      }
    })
  })

  it('will receive right props', () => {
    const props = mapStateToProps(store.getState())
    expect(props).toMatchSnapshot()
  })

  it('will receive right actions', () => {
    const actions = mapDispatchToProps(x => x)
    expect(actions).toMatchSnapshot()
  })
})
