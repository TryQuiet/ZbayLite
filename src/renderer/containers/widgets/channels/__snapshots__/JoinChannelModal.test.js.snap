// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JoinChannelModal will receive right actions 1`] = `
Object {
  "joinChannel": [Function],
  "showNotification": [Function],
}
`;

exports[`JoinChannelModal will receive right props 1`] = `
Object {
  "publicChannels": Object {},
  "users": Object {},
}
`;
